os/arch/xtensa/src/esp32/esp32_i2c.c

```diff
@@ -706,36 +757,30 @@ static void i2c_esp32_send_read_cmd(struct esp32_i2c_priv_s *priv, int addr_byte
 	 * slave device.  Divide the read command in two segments as
 	 * recommended by the ESP32 Technical Reference Manual.
 	 */
-	if (msg->length - to_read <= 1) {
+	else if (msg->length == 1) {
 		/* Read the last byte and explicitly ask for an
 		 * acknowledgment.
 		 */
 		*cmd++ = (struct i2c_esp32_cmd) {
 			.opcode = I2C_ESP32_OP_READ,
 			.num_bytes = 1,
-			.ack_val = true,
+			.ack_val = 1,
 		};
 
 		/* Account for the `msg.len - 1` when clamping
 		 * transmission length to FIFO buffer size.
 		 */
 		to_read++;
-		/*No more msg to excute */
-		if (I2C_M_NORESTART != (msg->flags & I2C_M_NORESTART)) {
-			wait_cmd = cmd;
-			*cmd++ = (struct i2c_esp32_cmd) {
-				.opcode = I2C_ESP32_OP_STOP
-			};
-		}
-	}
-	if (!wait_cmd) {
-		*cmd++ = (struct i2c_esp32_cmd) {
-			.opcode = I2C_ESP32_OP_END
-		};
 	}
 
-	priv->last_data_count = to_read;
+	*cmd++ = (struct i2c_esp32_cmd) {
+		.opcode = I2C_ESP32_OP_END
+	};
+
+	I2C[priv->i2c_num]->int_ena.end_detect = 1;
 
+	priv->last_data_count = to_read;
+	return to_read;
```

Comment:
For this, how about change return type of 'i2c_esp32_send_read_cmd' to 'uint32_t' ?

Reply:
It will return "ESP32_I2C_READ_ERROR", which is " -1" for error code.